### PR TITLE
Fix/migration docs and logging

### LINKED ISSUE #2202: Update Migrate CLI documentation to say it's resumable
It wasn't clear from the documentation that the migrate CLI can be resumed. It required code inspection to confirm.

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**: Remediation for internal PDS-54990. Fixes #2202, #2204, and #2228.

**Implementation Description (bullets)**: Docs tickets: added a line of docs
Logging ticket (trickier): added a method RangeMigrator.checkStatus, and called it before creating the migration tasks.

The migration tasks are performed concurrently, so it was necessary to do the "where are we up to" check outside of the migrate method.

I noticed that `TransactionRangeMigrator` and `KvsRangeMigrator` are almost identical, but declined (with a heavy heart) to refactor that in this PR, to avoid scope creep.

**Concerns (what feedback would you like?)**:
* I haven't actually tested this, and in general, our migration testing story is pretty non-existent as far as I can tell. Is there a reasonable way of testing? @tmgordeeva may know.
* This is technically a breaking change, as a method was added to the `RangeMigrators` interface. However, this interface isn't used outside of `atlasdb-client`. I considered adding something to the checkpointer itself (e.g. have `GeneralTaskCheckpointer.createCheckpoints` return a bool to say whether it actually created checkpoints, and then have a method `getStatus` that would do what the new `RangeMigrator` method does. However, this would _actually_ have been a breaking change in our large internal product, which has a different `AbstractTaskCheckpointer` implementation.
* Copy-pasta. Should I submit a follow-up PR to clean up this rarely-touched area of the codebase?

**Where should we start reviewing?**: TableMigrator.java

**Priority (whenever / two weeks / yesterday)**: This week, but shouldn't merge until after the 0.53.0 release.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2241)

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/KvsRangeMigrator.java [hunk 1]
```diff
@@ -78,6 +80,32 @@
...
+    private void checkStatus(Transaction tx, int numRangeBoundaries) {
+        for (int rangeId = 0; rangeId < numRangeBoundaries; rangeId++) {
+            byte[] checkpoint = getCheckpoint(rangeId, tx);
+            if (checkpoint != null) {
+                log.info("({}/{}) Migration from table {} to table {} will start/resume at {}",
+                        rangeId,
+                        numRangeBoundaries,
+                        srcTable,
+                        destTable,
+                        PtBytes.encodeHexString(checkpoint)
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/KvsRangeMigrator.java]: what is this value?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/TableMigrator.java [hunk 2]
```diff
@@ -97,6 +97,9 @@ public void migrate() {
         }
         checkpointer.createCheckpoints(srcTable.getQualifiedName(), boundaryById);
 
+        // Look up the checkpoints and log start point (or done)
+        rangeMigrator.checkStatus(rangeBoundaries.size());
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/TableMigrator.java]: should we call this `logStatus`?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/KvsRangeMigrator.java [hunk 3]
```diff
@@ -78,6 +80,32 @@
         this.rowTransform = rowTransform;
     }
 
+    // Report (a) completion or (b) start row. Intended to be called before migration has started.
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/KvsRangeMigrator.java]: should probably move these docs to the interface

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/TableMigrator.java (hunk 2 above)
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/TableMigrator.java]: Sounds good, will rename.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/KvsRangeMigrator.java (hunk 1 above)
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/KvsRangeMigrator.java]: This is the row key from which the migration will resume, as stored in the checkpoints table.

From the [javadoc](https://github.com/palantir/atlasdb/blob/0.53.0/atlasdb-client/src/main/java/com/palantir/atlasdb/schema/AbstractTaskCheckpointer.java#L48): " Get the next row name for this range. Returns null if the checkpoint is done."

**Comment by gsheasby** [comment]
@nziebart Applied the changes suggested. Ack, I can do the start/commit timestamp logging separately. I only see one place in `KeyValueServiceMigrator` where we grab a timestamp, so I can just log that timestamp.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/RangeMigrator.java [hunk 4]
```diff
@@ -18,5 +18,14 @@
...
 
 public interface RangeMigrator {
+
+    /**
+     * Report (a) completion or (b) start row.
+     * Intended to be called before starting or resuming migration.
+     *
+     * @param size the number of separate ranges in the current table
+     */
+    void logStatus(int size);
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/RangeMigrator.java]: nit: rename `size` to `numRangeBoundaries `

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/TransactionRangeMigrator.java [hunk 5]
```diff
@@ -62,6 +68,31 @@
     }
 
     @Override
+    public void logStatus(int size) {
+        txManager.runTaskWithRetry((TransactionTask<Void, RuntimeException>) transaction -> {
```
Comment (nziebart) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/TransactionRangeMigrator.java]: do we need the cast here?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/schema/KeyValueServiceMigrator.java [hunk 6]
```diff
@@ -206,6 +206,8 @@ private void migrateTables(Set<TableReference> tables,
                                long migrationTimestamp,
                                ExecutorService executor,
                                GeneralTaskCheckpointer checkpointer) {
+        processMessage("Migrating tables at migrationTimestamp " + migrationTimestamp,
```
Comment (nziebart) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/schema/KeyValueServiceMigrator.java]: really wish this was named something better than `processMessage`.. is it logging this? anyway, not a blocker for this PR

Comment (gsheasby) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/schema/KeyValueServiceMigrator.java]: It sends the message to an `OutputPrinter`. There's a chain of `processMessage` methods - they could all be `printMessage`?

I suspected that we'd break `__LARGE_INTERNAL_PRODUCT` if we renamed them all, but turns out said behemoth does not call the method directly.

**Comment by gsheasby** [comment]
Should be good to go now.

**Pull request merged**